### DISCUSSION TITLE
Socket.IO v3 Release - feedback

### DISCUSSION THREAD

**Discussion by darrachequesne** [socketio/socket.io#3674] [General]
Related blog post: https://socket.io/blog/socket-io-3-release/

If you have any remarks/suggestions about this release, let's discuss about it here!

**Comment by szarkowicz** [comment]
@darrachequesne Great news on socket.io v3 release going live!!

In the section: **What Socket.IO brings:**  Doesnt socket.io give you a lot more features out fo the box, such as Namespaces, rooms, auto-reconnect, scalability with Redis, different types of broadcast options without having to write custom code / logic for all that? 

I looked into WS after you mentioned it an it seems like a lot more work to get that sort of functionality working or having to try and find 3rd party libraries for all that.

Am I missing something??

Thanks again for your hard work.

**Reply by darrachequesne** [reply]
>  Doesnt socket.io give you a lot more features out fo the box, such as Namespaces, rooms, auto-reconnect, scalability with Redis, different types of broadcast options without having to write custom code / logic for all that?

You're right, I've added the auto-reconnection feature that was missing from the list. The other items should already be there, could you please check? Is there something wrong with the wording?

**Reply by szarkowicz** [reply]
Okay great - yea I think I just read through it a little fast and my phone and the way it looked like What socket.io brings looked only like fallback support. Sorry for the confusion.

**Comment by david-fong** [comment]
Hello! I've just switched over to v3. Thanks for your hard work!

There are some typing exports that I was able to use in v2, but currently can't use in v3.0.0. Specifically, I'm wondering if you are open to adding `export type { Namespace } from "./namespace"; export type { Client} from "./client";` in the [index file](https://github.com/socketio/socket.io/blob/master/lib/index.ts#L19) of the serverside implementation.

It may be helpful to link the [migration guide](https://socket.io/docs/migrating-from-2-x-to-3-0/) in the links section of [the GitHub release for v3.0.0](https://github.com/socketio/socket.io-client/releases/tag/3.0.0)

**Reply by darrachequesne** [reply]
> There are some typing exports that I was able to use in v2, but currently can't use in v3.0.0

Well, why not. Out of curiosity, how are you using those exports?

> It may be helpful to link the migration guide in the links section of the GitHub release for v3.0.0

Done!

**Reply by david-fong** [reply]
Thanks, Damien!

> Well, why not. Out of curiosity, how are you using those exports?

I'm working on a multiplayer game. Each client will connect to three kinds of namespaces. There's a namespace that tracks existing groups, and for each group, there is a namespace for the lobby (with events such as players joining and leaving the group), and a namespace for events that occur while a game is in session.

[My Group class](https://github.com/david-fong/snakey3/blob/dev/src/server/Group.ts#L22) encapsulates the group's lobby namespace, and [my game class](https://github.com/david-fong/snakey3/blob/dev/src/server/ServerGame.ts#L30) encapsulates the group's in-game events namespace. I import socket.io like so, `import type * as io from "socket.io";`, and I used to be able to declare fields like `public readonly namespace: io.Namespace`, but now I do `public readonly namespace: io.Socket["nsp"]`, which doesn't really bother me, but I wish the old way still worked.

[Here's some of my socket.io-client connection code](https://github.com/david-fong/snakey3/blob/dev/src/client/SkSockets.ts).

**Reply by darrachequesne** [reply]
The exports were added in [3.0.1](https://github.com/socketio/socket.io/releases/tag/3.0.1) (commit: https://github.com/socketio/socket.io/commit/f62f180edafdd56d8a8a277e092bc66df0c5f07f).

**Reply by david-fong** [reply]
Awesome! Thanks, Damien.

**Comment by jsarenik** [comment]
It's always good to see life! Thank you for this release! All the best!

**Comment by achimnohl** [comment]
Great to see this project is alive. I have been using it and relying on it for years in multiple projects. 
You stated "the client implementations in other languages will be updated as well". It would be great if the CPP client could be updated as it is still based on 1.x and AFAIK not compatible with >=2. We now have a use case where we need to talk to a CPP client. I understand that this may not be a very popular requirement though. Big thanks to the Socket.IO team for your hard work!

**Reply by darrachequesne** [reply]
This is indeed on my TODO list.

**Reply by darrachequesne** [reply]
The C++ client is now compatible with Socket.IO v3: https://github.com/socketio/socket.io-client-cpp/releases/tag/3.0.0

**Comment by hyperknot** [comment]
There are two small things I found related to v3 and docs:
1. https://github.com/socketio/socket.io/issues/3692
2. https://github.com/socketio/socket.io-client/issues/1401

**Comment by n1ru4l** [comment]
Hey there, love the new update. I recently wrote a library for doing GraphQL over Socket.io and updated it to v3, I think it could be a nice addition for the documentation! https://github.com/n1ru4l/graphql-live-query/tree/main/packages/socket-io-graphql-server

**Comment by spiritbroski** [comment]
hello guys  congrats on the release, i just want to ask that the java client doesnt work with this release the latest release that can work with this release is 2.3.0 the 3.0.0 up release doesnt work with java client so i hope somebody can fix that thanks

**Comment by Sudeep-1** [comment]
Hi guys, Great to hear about the new version. Is the new version 3.0 available for android and IOS?

**Reply by jsarenik** [reply]
Do you want to run the server part on Android/iOS? The client part requires only a browser with JavaScript (be it Google Chrome, Firefox, Safari, …) so yes, the client is available (works) on both Android and iOS.

**Comment by spiritbroski** [comment]
no as of now the socket io client for java still doesnt work with this
version

**Reply by Sudeep-1** [reply]
oh...When can we expect it...?

**Reply by GingerAdonis** [reply]
It's available right now. :)

**Comment by spiritbroski** [comment]
i dont know im not the maintainer looks like the maintainer busy with
something else

**Reply by Sudeep-1** [reply]
Hmm..wish the new version was available for android and IOS. Was planning to use socket.io in my app. I may need to go for OKHTTP then.

**Comment by spiritbroski** [comment]
yup go for alternative for now because i think the new version still along
way to go the creator of this socket.io is also the creator of vercel so i
think they are focused to that too now so in order for this to be
maintained i think its complicated

**Reply by darrachequesne** [reply]
Hi! Guillermo is indeed no longer actively maintaining the project, as he is (understandably) quite busy with Vercel.

The Socket.IO project is still maintained though, we plan to update the Java client in the following weeks.

**Reply by darrachequesne** [reply]
Update: the version [2.0.0](https://github.com/socketio/socket.io-client-java/releases/tag/socket.io-client-2.0.0) of the Java client is now compatible with Socket.IO v3.

The migration guide can be found here: http://socketio.github.io/socket.io-client-java/migrating_from_1_x.html

**Comment by spiritbroski** [comment]
wow thank you man looking forward to it

**Comment by spiritbroski** [comment]
that is not the issue i already got it working fine with socket.io java
client version 1.0.0 with socket.io-server version 2.3.0 but with socket io
server version 3.0.0 it doesnt work,then the maintainer recently said that
the new version of java client that works with 3.0.0 is still in the works
so thats why i still stick with the version 2.3.0

**Comment by momakes2** [comment]
Just love the work you're putting into this, bravo.

**Comment by RosenTomov** [comment]
@darrachequesne 
I've noticed a weird issue, where it would randomly not match the namespace with regex.
For example, with namespace "/1", if it doesn't connect, I have to restart the server until it does.

`io.of(/^\/(\d+)$/).use(socketMiddleware).on('connection', socketFunctions(io)); // randomly works`

`io.of("/1").use(socketMiddleware).on('connection', socketFunctions(io)); // always works`

**Comment by spiritbroski** [comment]
Awesome, thanks for the update!
